add ability to show contents of service fdstore + teach systemd-notify passing fds into the fdstore

poettering opened the pull request.

dtardon commented on src/core/dbus-service.c:
```diff
@@ -218,6 +219,72 @@ int bus_service_method_mount_image(sd_bus_message *message, void *userdata, sd_b
+                /* glibc implies O_LARGEFILE everywhere on 64bit off_t builds, but forgets to hide it away on
+                 * F_GETFL, but provides no definition to check for that. Let's mask the flag away manually,
+                 * to not confuse clients. */
+                flags &= RAW_O_LARGEFILE;
```
Missing negation?

poettering replied:
ouch ouch ouch

DaanDeMeyer commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -2548,6 +2557,7 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
                  in  b read_only,
                  in  b mkdir,
                  in  a(ss) options);
+      DumpFileDescriptorStore(out a(suuutuusu) entries);
```
Shouldn't we do this via a memfd containing JSON? At least that way if we ever need to add something in the future we can do so without adding a new method call.

poettering replied:
We can certainly add that one day if needed. But note that the dump code that uses memfd uses that for size reasons (you cannot pass arbitrarily large dumps via dbus native proto), not for a reaspm pf "we want something less rigid than the dbus type system". The size thing is a hard issue, we ran into this IRL, so we added the memfd as workaround, and it's not more than that because inherently incompatible with remoting operation via ssh or such. If we really run into issues with the message sizes for the fdstore too, then we can go down the memfd route too, as an alternative. But I'd be really careful with the other issue: yeah dbus type system is inflexible, but as long as we do use dbus we should buy into it, otherwise we degrade dbus to just a noise envelope on AF_UNIX, which brings no benefits over just using AF_UNIX directly....

I am not too concerned that we want a lot of extensibility here. fstat() hasn't changed in 40 years after all, and statx() doesn't bring much relevant to the table for this case here either. /proc/self/fd is pretty stable too. Hence I wouldn't know what to add here even if we wanted to.

Here's another way to see this: we probably should extend the existing Dump() calls to also show the fdstore – we currently don't and that really should be addressed. Dump() already does memfds, and is not bound by the dbus type system.

Hence, let's keep DumpFileDescriptorStore() separate from that. Let's have the structured, client-friendly, high-level DumpFileDescriptorStore() and then (in a separate PR) also extend the messier, low-level, free-form, memfd-based regular Dump() to show the fdstore, too.

DaanDeMeyer commented on src/analyze/analyze-fdstore.c:
```diff
@@ -0,0 +1,109 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to mangle name '%s': %m", arg);
+
+        r = bus_message_new_method_call(bus, &m, bus_systemd_mgr, "DumpUnitFileDescriptorStore");
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_append(m, "s", unit);
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_call(bus, m, 0, &error, &reply);
+        if (r < 0)
+                return log_error_errno(r, "Failed to call DumpUnitFileDescriptorStore: %s",
+                                       bus_error_message(&error, r));
```
`bus_call_method()`?

DaanDeMeyer commented on src/shared/fdset.c:
```diff
@@ -23,27 +24,29 @@ FDSet *fdset_new(void) {
-int fdset_new_array(FDSet **ret, const int *fds, size_t n_fds) {
-        size_t i;
-        FDSet *s;
+static inline void fdset_shallow_freep(FDSet **s) {
```
Shouldn't we rename `fdset_free()` to `fdset_free_free()` and name this one `fdset_free()`, to mimick what we do with hashmaps and such?

poettering replied:
hmm, this is an internal helper, and only called in exceptional cases. And I think this is unlikely to change, given it really just exists to make our exported functions somewhat "atomic": either they succeed or they fail, but they won't take possesion of parameters on failure. Outside of the internal implementaiton we won't need that however.

I think "fdset_free()" as the primary exported function should use the generic name, and the internal, exceptional case get the special name.

bluca commented on src/core/dbus-service.c:
```diff
@@ -218,6 +219,72 @@ int bus_service_method_mount_image(sd_bus_message *message, void *userdata, sd_b
+
+        assert(message);
+
+        r = mac_selinux_unit_access_check(UNIT(s), message, "status", error);
```
I wonder if "status" isn't too lax for this. After all, this information used to be private. Are we showing only metadata about the FDs, or is there any way for an unprivileged user to use it to derive information about the content of the FDs too?

poettering replied:
i am not aware of any more appropriate selinux action for this. we can't just make up our own. the thing is that selinux policy upstream never responds to our requests about reworking the action names, hence I doubt we should innovate here.

poettering replied:
note that "systemd-analyze dump" is protected by the same selinux verb. We don't need to be more catholic than the pope on this hence.

poettering replied:
i think it's fine to list the file names. It would be problem if we'd show the contents of the files though

bluca replied:
> note that "systemd-analyze dump" is protected by the same selinux verb. We don't need to be more catholic than the pope on this hence.

Does that show the file names and other metadata associated with stored FDs?

> i am not aware of any more appropriate selinux action for this. we can't just make up our own. the thing is that selinux policy upstream never responds to our requests about reworking the action names, hence I doubt we should innovate here.

I can worry about that part, if needed. If we are saying that the filenames and the other things showed here are not privileged that's fine, but to be extra clear it would be worth to mention in the fdstore manpage that unprivileged users will generally be able to see that info. We don't want anybody to be surprised.

poettering replied:
the generic dbus dump calls do not show fdstore contnets right now. But we should change that. They should. see other comments about that.

I'll update the man page to make clear that contents of fds int he fdstore is considered sensitive, but not the fd metadata.

bluca commented on src/core/dbus-service.c:
```diff
@@ -218,6 +219,72 @@ int bus_service_method_mount_image(sd_bus_message *message, void *userdata, sd_b
+                return r;
+
+        if (s->n_fd_store_max == 0 && s->n_fd_store == 0)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "File descriptor store not enabled for %s.", UNIT(s)->id);
```
Should this simply return an empty array? After all it's not really an error, it's just that there's no info to show

poettering replied:
we return a empty array if the fdstore is enabled but nothing in there. I think there's value in allowing clients to distinguish the cases "fdstore disabled" from "fdstore enabled but empty". The former gets this error, the latter an empty array.

Note the check checks if the limit and fill level is both zero. Only if both are zero we'll return the error. this is to cover cases where a service is started with fdstore on, receives a bunch of fds, and then a daemon reload is issued and the unit changed to disallow an fdstore. In such a case we should still show the fdstore that was acquired, even if now we wouldn't allow that anymore.

bluca replied:
I'm not sure - "invalid arguments" to be sounds like the input was bogus, but it is clearly not the case here, it's simply that an option was disabled. When receiving `SD_BUS_ERROR_INVALID_ARGS`, my immediate reaction is to suspect there's a coding error in the client and it's sending bogus inputs.

DaanDeMeyer commented on src/shared/fdset.c:
```diff
@@ -192,24 +192,39 @@ int fdset_new_listen_fds(FDSet **ret, bool unset) {
-int fdset_close_others(FDSet *fds) {
-        _cleanup_free_ int *a = NULL;
-        size_t j = 0, m;
+int fdset_get_array(FDSet *fds, int **ret_array) {
```
`fdset_to_array()`?

`ret_array` can just be named `ret`

bluca commented on src/analyze/analyze-fdstore.c:
```diff
@@ -0,0 +1,109 @@
+        _cleanup_(table_unrefp) Table *table = NULL;
+        _cleanup_free_ char *unit = NULL;
+        int r;
+
```
assert(bus)

DaanDeMeyer commented on man/systemd-notify.xml:
```diff
@@ -191,6 +191,24 @@
+        <listitem><para>Send a file descriptor along with the notification message. This is useful when
+        invoked in services that have the <varname>FileDescriptorStoreMax=</varname> setting enabled, see
+        <citerefentry><refentrytitle>systemd.service</refentrytitle><manvolnum>5</manvolnum></citerefentry>
+        for details. The specified file descriptor must be passed to <command>systemd-notify</command> when
```
Maybe mention how this can be done in bash? https://github.com/containers/bubblewrap/blob/main/demos/bubblewrap-shell.sh is a good example

bluca replied:
yeah, this is obscure enough to warrant a specific mention and example here

DaanDeMeyer commented on src/notify/notify.c:
```diff
@@ -198,6 +211,62 @@ static int parse_argv(int argc, char *argv[]) {
+                                return log_error_errno(SYNTHETIC_ERRNO(ERANGE), "File descriptor can't be negative: %i", fdnr);
+
+                        if (!passed) {
+                                /* Take possession of all assed fds */
```
Typo

bluca replied:
...or is it!

bluca commented on src/analyze/analyze-fdstore.c:
```diff
@@ -0,0 +1,109 @@
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_call(bus, m, 0, &error, &reply);
```
`bus_call_method(bus, bus_systemd_mgr, "DumpUnitFileDescriptorStore", &error, &reply, "s", unit)` to shorten this a bit

DaanDeMeyer commented on src/notify/notify.c:
```diff
@@ -198,6 +211,62 @@ static int parse_argv(int argc, char *argv[]) {
+                                        return log_error_errno(owned_fd, "Specified file descriptor '%i' not passed or specified more than once: %m", fdnr);
+                        }
+
+                        if (!arg_fds) {
+                                arg_fds = fdset_new();
+                                if (!arg_fds)
+                                        return log_oom();
+                        }
```
Add an `ensure_allocated` helper like we have for regular sets?

bluca commented on src/analyze/analyze.c:
```diff
@@ -229,6 +230,7 @@ static int help(int argc, char *argv[], void *userdata) {
                "  security [UNIT...]         Analyze security of unit\n"
                "  inspect-elf FILE...        Parse and print ELF package metadata\n"
                "  malloc [D-BUS SERVICE...]  Dump malloc stats of a D-Bus service\n"
+               "  fdstore [SERVICE...]       Show file descriptor store contents of service\n"
```
at least one service is mandatory, so `fdstore SERVICE...` no?

poettering replied:
true

DaanDeMeyer commented on src/notify/notify.c:
```diff
@@ -198,6 +211,62 @@ static int parse_argv(int argc, char *argv[]) {
+                                        return log_oom();
+                        }
+
+                        r = fdset_put(arg_fds, owned_fd);
```
`Add fdset_consume()`?

DaanDeMeyer reviewed (changes requested):
Looks good but I think we should dump via a memfd of JSON

bluca commented on src/shared/fdset.c:
```diff
@@ -192,24 +192,39 @@ int fdset_new_listen_fds(FDSet **ret, bool unset) {
-                a = new(int, m);
-                if (!a)
-                        return -ENOMEM;
+        if (m > INT_MAX) /* We want to be able to return an "int" */
```
shouldn't this check if it's zero too?

poettering replied:
why?

this is just a paranoid check to ensure that `m` (which is unsigned) will fit into an int too, since that's how we'd ultimately like to return it.

bluca replied:
The previous code did - it allocated `a` only if `m > 0`. What happens if `m` is 0 here? You get a 1-sized array, that is actually empty? Isn't that a bit weird?

poettering commented:
Addressed all points, except for the memfd thing (see comments about that, i.e. i think we should include fdstore info in the regular dump, to address this). 

I also didn't add the fset_ensure_consume() thing yet, as it would only have a singler user. Let's add that once we have a second one.

bluca commented on src/analyze/analyze-fdstore.c:
```diff
@@ -0,0 +1,111 @@
+                return log_error_errno(r, "Failed to call DumpUnitFileDescriptorStore: %s",
+                                       bus_error_message(&error, r));
+
+
```
double newline

poettering replied:
fixed.
